singularity-opac/photons/epbremsstrahlung_opacity_photons.hpp

```diff
@@ -0,0 +1,191 @@
+                       Real *lambda = nullptr) const {
+    return 4 * M_PI * EmissivityPerNuOmega(rho, temp, nu, lambda);
+  }
+
+  template <typename FrequencyIndexer, typename DataIndexer>
+  PORTABLE_INLINE_FUNCTION void
+  EmissivityPerNu(const Real rho, const Real temp, FrequencyIndexer &nu_bins,
+                  DataIndexer &coeffs, const int nbins,
+                  Real *lambda = nullptr) const {
+    for (int i = 0; i < nbins; ++i) {
+      coeffs[i] = EmissivityPerNu(rho, temp, nu_bins[i], lambda);
+    }
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real Emissivity(const Real rho, const Real temp,
+                  Real *lambda = nullptr) const {
+
+    const Real thetaE = pc::kb * temp / (pc::me * pc::c * pc::c);
+    const Real ne = rho / mmw_;
+    const Real ni = ne;
+    return 4. * M_PI * pc::kb * temp / pc::h * prefac_ / std::sqrt(thetaE) *
+           ne * ni * gff_;
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real NumberEmissivity(const Real rho, const Real temp,
+                        Real *lambda = nullptr) const {
+    // Infrared catastrophe
+    return std::numeric_limits<Real>::infinity();
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real ThermalDistributionOfTNu(const Real temp, const Real nu,
+                                Real *lambda = nullptr) const {
+    return dist_.ThermalDistributionOfTNu(temp, nu, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real DThermalDistributionOfTNuDT(const Real temp, const Real nu,
+                                   Real *lambda = nullptr) const {
+    return dist_.DThermalDistributionOfTNuDT(temp, nu, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real ThermalDistributionOfT(const Real temp, Real *lambda = nullptr) const {
+    return dist_.ThermalDistributionOfT(temp, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION Real
+  ThermalNumberDistributionOfT(const Real temp, Real *lambda = nullptr) const {
+    return dist_.ThermalNumberDistributionOfT(temp, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real EnergyDensityFromTemperature(const Real temp,
+                                    Real *lambda = nullptr) const {
+    return dist_.EnergyDensityFromTemperature(temp, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real TemperatureFromEnergyDensity(const Real er,
+                                    Real *lambda = nullptr) const {
+    return dist_.TemperatureFromEnergyDensity(er, lambda);
+  }
+
+  PORTABLE_INLINE_FUNCTION
+  Real NumberDensityFromTemperature(const Real temp,
+                                    Real *lambda = nullptr) const {
+    return dist_.NumberDensityFromTemperature(temp, lambda);
+  }
+
+ private:
+  Real mass_ion_;
+  PlanckDistribution<pc> dist_;
+  static constexpr Real mmw_ =
+      (pc::mp + pc::me) /
+      (2. * pc::mp); // Neutral fully ionized electron-proton gas
+  static constexpr Real gff_ = 1.2;
+  static constexpr Real prefac_ = 8. * std::pow(pc::qe, 6) /
+                                  (3. * std::pow(pc::me * pc::c * pc::c, 2)) *
```

Comment:
didn't know `std::pow` was constexpr... nice.

Reply:
OK just removed the `static constexpr` decorators. In the future if we do want expressions like this to be `constexpr` we could add a dependency to [Sprout](https://github.com/bolero-MURAKAMI/Sprout)

Reply:
these are also integer powers, so we could just write them out, or even implement our own constexpr pow. Something along the lines of the code here:
https://stackoverflow.com/questions/17719674/c11-fast-constexpr-integer-powers

Reply:
From stack overflow:
```C++
#include <cstdint>

constexpr int64_t ipow(int64_t base, int exp, int64_t result = 1) {
  return exp < 1 ? result : ipow(base*base, exp/2, (exp % 2) ? result*base : result);
}

int64_t foo(int64_t base, int exp) {
  return ipow(base, exp);
}

```

Reply:
Something like... (checked in godbolt):
```C++
constexpr Real mypow(Real x, std::size_t p) {
    if (p == 0) return 1;
    if (p == 1) return x;
    if ((p % 2) == 0) return mypow(x, p/2)*mypow(x, p/2);
    return mypow(x, p - 1)*x;
}
```